### PR TITLE
pkp/pkp-lib#7285 Create PHPMailer transport

### PR DESCRIPTION
**Pull request by Vitaliy-1**

### CODE REVIEW

File: classes/mail/transport/PHPMailTransport.php [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+class PHPMailTransport implements TransportInterface
+{
+    public const MAIL_WRAP = 76;
+
+    /**
+     * @inheritDoc
+     */
+    public function send(RawMessage $symfonyMessage, Envelope $envelope = null): ?SentMessage
+    {
+        if (!is_a($symfonyMessage, Email::class)) {
```
Comment (jonasraoni) [inline classes/mail/transport/PHPMailTransport.php:38]: Maybe `$symfonyMessage instanceof Email`?

File: classes/mail/transport/PHPMailTransport.php [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+use ReflectionObject;
+
+class PHPMailTransport implements TransportInterface
+{
+    public const MAIL_WRAP = 76;
+
+    /**
+     * @inheritDoc
+     */
+    public function send(RawMessage $symfonyMessage, Envelope $envelope = null): ?SentMessage
```
Comment (jonasraoni) [inline classes/mail/transport/PHPMailTransport.php:36]: I think it should be using `?Envelope`

File: classes/mail/transport/PHPMailTransport.php [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        $fromAddresses = $symfonyMessage->getFrom();
+        if (!empty($fromAddresses)) {
+            $f = current($fromAddresses);
+            if (Config::getVar('email', 'force_default_envelope_sender') && Config::getVar('email', 'default_envelope_sender') && Config::getVar('email', 'force_dmarc_compliant_from')) {
+                /* If a DMARC compliant RFC5322.From was requested we need to promote the original RFC5322.From into a Reply-to header
+                 * and then munge the RFC5322.From */
+                $alreadyExists = false;
+                foreach ($symfonyMessage->getReplyTo() as $r) {
+                    if ($r->getAddress() === $f->getAddress()) {
+                        $alreadyExists = true;
```
Comment (jonasraoni) [inline classes/mail/transport/PHPMailTransport.php:85]: I think you can add a break here, or perhaps use the https://laravel.com/docs/9.x/helpers#method-array-first

File: classes/mail/transport/PHPMailTransport.php [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                    $patterns = ['#%n#', '#%s#'];
+                    $replacements = [$f->getName(), $site->getLocalizedTitle()];
+                    $fromName = preg_replace($patterns, $replacements, Config::getVar('email', 'dmarc_compliant_from_displayname'));
+                } else {
+                    $fromName = '';
+                }
+                $fromEmail = Config::getVar('email', 'default_envelope_sender');
+            }
+            // this sets both the envelope sender (RFC5321.MailFrom) and the From: header (RFC5322.From)
+            $mailer->SetFrom($fromEmail ?? $f->getAddress(), $fromName ?? $f->getName());
```
Comment (jonasraoni) [inline classes/mail/transport/PHPMailTransport.php:105]: As `$fromName` might be an empty string, maybe here it should be `$fromName ?: $f->getName()`?

File: classes/mail/transport/PHPMailTransport.php (hunk 2 above)
Comment (jonasraoni) [reply classes/mail/transport/PHPMailTransport.php:36]: Ah, I just saw the `@inheritDoc`... If this is part of the interface, so be it 😁

**Review by jonasraoni — APPROVED** [review]

File: classes/mail/transport/PHPMailTransport.php (hunk 3 above)
Comment (Vitaliy-1) [reply classes/mail/transport/PHPMailTransport.php:85]: Yep, look like it

File: classes/mail/transport/PHPMailTransport.php (hunk 4 above)
Comment (Vitaliy-1) [reply classes/mail/transport/PHPMailTransport.php:105]: Missed that, thanks!

Comment (jonasraoni) [reply classes/mail/transport/PHPMailTransport.php]: Great, so just a comment/test could be added to prevent people like me from breaking things 😂

**Comment by asmecher** [comment]
Looks good, @Vitaliy-1, but I think using `phpmail` in various places rather than consistently using `phpmailer`, will generate some confusion about whether we're talking about PHPMailer the package or PHP's `mail()` function. I'd recommend changing out `phpmail` (and e.g. `PHPMailTransport`) for `phpmailer` / `PHPMailerTransport`. Could you make those changes and tag me on the issue? Then I'll play around with testing it locally and complete the review.

**Comment by Vitaliy-1** [comment]
@asmecher, done! Regarding DMARC stuff, I'm planning to remove it from here once I finish the work on the next issue.

**Comment by Vitaliy-1** [comment]
(updated because of failed tests)

File: classes/mail/transport/PHPMailerTransport.php [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+{
+    public const MAIL_WRAP = 76;
+
+    /**
+     * @inheritDoc
+     */
+    public function send(RawMessage $symfonyMessage, Envelope $envelope = null): ?SentMessage
+    {
+        if (!($symfonyMessage instanceof Email)) {
+            throw new Exception('Can\'t send raw message with phpmail');
```
Comment (asmecher) [inline classes/mail/transport/PHPMailerTransport.php:39]: (should be `phpmailer`)

File: classes/mail/transport/PHPMailerTransport.php [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        $mailer->Encoding = PHPMailer::ENCODING_BASE64;
+        $mailer->CharSet = $symfonyMessage->getHtmlCharset() ?? 'utf-8';
+        $mailer->XMailer = 'Public Knowledge Project Suite v3';
+        $mailer->WordWrap = static::MAIL_WRAP;
+
+        $request = Application::get()->getRequest();
+        $fromAddresses = $symfonyMessage->getFrom();
+        if (!empty($fromAddresses)) {
+            $f = current($fromAddresses);
+            if (Config::getVar('email', 'force_default_envelope_sender') && Config::getVar('email', 'default_envelope_sender') && Config::getVar('email', 'force_dmarc_compliant_from')) {
```
Comment (asmecher) [inline classes/mail/transport/PHPMailerTransport.php:79]: In the `config.TEMPLATE.inc.php` files for each application, could you note which options are supported by which transports?

**Review by asmecher — APPROVED** [review]
Just a couple trivial comments and these PRs are all ready to go, @Vitaliy-1. I'm still really pleased that this ended up being such a short/clean migration of our old email sending code to a new transport. Kudos!

File: classes/mail/transport/PHPMailerTransport.php (hunk 6 above)
Comment (Vitaliy-1) [reply classes/mail/transport/PHPMailerTransport.php:79]: My next PR implements this for all transports in general and it's going to be removed from here.

**Pull request merged**